File: frontend/src/services/api/company/jobs.ts
/**
 * Jobs (job posting) API service — company side.
 * Consumes the /company/jobs router. `request()` auto-attaches the Bearer token.
 *
 *   POST   /company/jobs                              create (draft, or live if publish=true)
 *   GET    /company/jobs[?status=]                    list own jobs
 *   GET    /company/jobs/{id}                         one job (with weights)
 *   PATCH  /company/jobs/{id}                         edit
 *   POST   /company/jobs/{id}/publish                 draft → live
 *   POST   /company/jobs/{id}/close                   → closed
 *   POST   /company/jobs/draft-description            Gemini JD draft
 *   GET    /company/jobs/options/colleges             college list for restricted visibility
 *   GET    /company/jobs/{jobId}/applicants[?sort_by=] ranked candidates
 *   GET    /company/jobs/{jobId}/applicants/{appId}   full job-scoped analysis
 *   PATCH  /company/jobs/{jobId}/applicants/{appId}/status
 */

import { request } from "../client";
import type {
  Application,
  ApplicationAnalysis,
  ApplicationAnalysisJson,
  ApplicationStatus,
  CollegeOption,
  Job,
  JobCreatePayload,
  JobDomain,
  JobEmploymentType,
  JobExperienceLevel,
  JobInterviewDuration,
  JobInterviewMode,
  JobStatus,
  JobUpdatePayload,
  JobVisibility,
  JobWeights,
  RankedApplicant,
} from "@/types/jobs";

type Raw = Record<string, unknown>;

// ── snake_case DB → camelCase TS ─────────────────────────────────────

function normalizeWeights(raw: Raw | null | undefined): JobWeights | null {
  if (!raw) return null;
  return {
    resumeWeight: Number(raw.resume_weight ?? 0),
    githubWeight: Number(raw.github_weight ?? 0),
    leetcodeWeight: Number(raw.leetcode_weight ?? 0),
    interviewWeight: Number(raw.interview_weight ?? 0),
    assessmentWeight: Number(raw.assessment_weight ?? 0),
  };
}

function toWeightsPayload(w: JobWeights) {
  return {
    resume_weight: w.resumeWeight,
    github_weight: w.githubWeight,
    leetcode_weight: w.leetcodeWeight,
    interview_weight: w.interviewWeight,
    assessment_weight: w.assessmentWeight,
  };
}

function normalizeJob(raw: Raw): Job {
  return {
    id: raw.id as string,
    companyId: raw.company_id as string,
    title: raw.title as string,
    description: raw.description as string,
    domain: raw.domain as JobDomain,
    experienceLevel: raw.experience_level as JobExperienceLevel,
    location: raw.location as string,
    openingsCount: Number(raw.openings_count ?? 1),
    deadline: raw.deadline as string,
    visibility: raw.visibility as JobVisibility,
    status: raw.status as JobStatus,
    skills: (raw.skills as string[]) ?? [],
    visibleCollegeIds: (raw.visible_college_ids as string[]) ?? [],
    weights: normalizeWeights(raw.weights as Raw | null),
    applicationCount: Number(raw.application_count ?? 0),
    employmentType: (raw.employment_type as JobEmploymentType) ?? "full-time",
    interviewDurationMinutes: Number(raw.interview_duration_minutes ?? 30) as JobInterviewDuration,
    interviewMode: (raw.interview_mode as JobInterviewMode) ?? "ai",
    isOnCampus: raw.is_on_campus === undefined ? true : Boolean(raw.is_on_campus),
    minCgpa: raw.min_cgpa == null ? null : Number(raw.min_cgpa),
    eligibleBranches: (raw.eligible_branches as string[]) ?? [],
    eligibleBatchYears: (raw.eligible_batch_years as number[]) ?? [],
    ctcMin: raw.ctc_min == null ? null : Number(raw.ctc_min),
    ctcMax: raw.ctc_max == null ? null : Number(raw.ctc_max),
    ctcCurrency: (raw.ctc_currency as string | null) ?? "INR",
    stipendMin: raw.stipend_min == null ? null : Number(raw.stipend_min),
    stipendMax: raw.stipend_max == null ? null : Number(raw.stipend_max),
    internshipDurationMonths:
      raw.internship_duration_months == null ? null : Number(raw.internship_duration_months),
    ppoCtcMin: raw.ppo_ctc_min == null ? null : Number(raw.ppo_ctc_min),
    ppoCtcMax: raw.ppo_ctc_max == null ? null : Number(raw.ppo_ctc_max),
    perks: (raw.perks as string[]) ?? [],
    createdAt: raw.created_at as string,
    updatedAt: raw.updated_at as string,
  };
}

function normalizeApplication(raw: Raw): Application {
  return {
    id: raw.id as string,
    studentId: raw.student_id as string,
    jobId: raw.job_id as string,
    companyId: raw.company_id as string,
    status: raw.status as ApplicationStatus,
    appliedAt: raw.applied_at as string,
    updatedAt: raw.updated_at as string,
    jobTitle: (raw.job_title as string | null) ?? null,
    companyName: (raw.company_name as string | null) ?? null,
    placementProbability: (raw.placement_probability as number | null) ?? null,
  };
}

function normalizeAnalysis(raw: Raw): ApplicationAnalysis {
  return {
    id: raw.id as string,
    applicationId: raw.application_id as string,
    studentId: raw.student_id as string,
    jobId: raw.job_id as string,
    companyId: raw.company_id as string,
    analysisJson: raw.analysis_json as ApplicationAnalysisJson,
    placementProbability: Number(raw.placement_probability),
    weightedComposite: Number(raw.weighted_composite),
    weightsSnapshot: raw.weights_snapshot as ApplicationAnalysis["weightsSnapshot"],
    generatedAt: raw.generated_at as string,
  };
}

function normalizeRanked(raw: Raw): RankedApplicant {
  return {
    applicationId: raw.application_id as string,
    studentId: raw.student_id as string,
    studentName: (raw.student_name as string | null) ?? null,
    studentEmail: (raw.student_email as string | null) ?? null,
    status: raw.status as ApplicationStatus,
    appliedAt: raw.applied_at as string,
    placementProbability: (raw.placement_probability as number | null) ?? null,
    weightedComposite: (raw.weighted_composite as number | null) ?? null,
    hasAnalysis: Boolean(raw.has_analysis),
  };
}

function toJobPayload(p: JobCreatePayload | JobUpdatePayload) {
  const body: Record<string, unknown> = {};
  if (p.title !== undefined) body.title = p.title;
  if (p.description !== undefined) body.description = p.description;
  if (p.domain !== undefined) body.domain = p.domain;
  if (p.experienceLevel !== undefined) body.experience_level = p.experienceLevel;
  if (p.location !== undefined) body.location = p.location;
  if (p.openingsCount !== undefined) body.openings_count = p.openingsCount;
  if (p.deadline !== undefined) body.deadline = p.deadline;
  if (p.visibility !== undefined) body.visibility = p.visibility;
  if (p.skills !== undefined) body.skills = p.skills;
  if (p.visibleCollegeIds !== undefined) body.visible_college_ids = p.visibleCollegeIds;
  if (p.weights !== undefined && p.weights) body.weights = toWeightsPayload(p.weights);
  if ("publish" in p && p.publish !== undefined) body.publish = p.publish;
  // Eligibility & offer detail
  if (p.employmentType !== undefined) body.employment_type = p.employmentType;
  if (p.interviewDurationMinutes !== undefined)
    body.interview_duration_minutes = p.interviewDurationMinutes;
  if (p.interviewMode !== undefined) body.interview_mode = p.interviewMode;
  if (p.isOnCampus !== undefined) body.is_on_campus = p.isOnCampus;
  if (p.minCgpa !== undefined) body.min_cgpa = p.minCgpa;
  if (p.eligibleBranches !== undefined) body.eligible_branches = p.eligibleBranches;
  if (p.eligibleBatchYears !== undefined) body.eligible_batch_years = p.eligibleBatchYears;
  if (p.ctcMin !== undefined) body.ctc_min = p.ctcMin;
  if (p.ctcMax !== undefined) body.ctc_max = p.ctcMax;
  if (p.ctcCurrency !== undefined) body.ctc_currency = p.ctcCurrency;
  if (p.stipendMin !== undefined) body.stipend_min = p.stipendMin;
  if (p.stipendMax !== undefined) body.stipend_max = p.stipendMax;
  if (p.internshipDurationMonths !== undefined)
    body.internship_duration_months = p.internshipDurationMonths;
  if (p.ppoCtcMin !== undefined) body.ppo_ctc_min = p.ppoCtcMin;
  if (p.ppoCtcMax !== undefined) body.ppo_ctc_max = p.ppoCtcMax;
  if (p.perks !== undefined) body.perks = p.perks;
  return body;
}

// ── Service ─────────────────────────────────────────────────────────

export const jobsService = {
  create: (payload: JobCreatePayload): Promise<Job> =>
    request<Raw>("/company/jobs", {
      method: "POST",
      body: toJobPayload(payload),
    }).then(normalizeJob),

  list: (status?: JobStatus): Promise<Job[]> =>
    request<Raw[]>(
      status ? `/company/jobs?status=${encodeURIComponent(status)}` : "/company/jobs",
    ).then((rows) => rows.map(normalizeJob)),

  getById: (jobId: string): Promise<Job> =>
    request<Raw>(`/company/jobs/${jobId}`).then(normalizeJob),

  update: (jobId: string, payload: JobUpdatePayload): Promise<Job> =>
    request<Raw>(`/company/jobs/${jobId}`, {
      method: "PATCH",
      body: toJobPayload(payload),
    }).then(normalizeJob),

  publish: (jobId: string): Promise<Job> =>
    request<Raw>(`/company/jobs/${jobId}/publish`, { method: "POST" }).then(normalizeJob),

  close: (jobId: string): Promise<Job> =>
    request<Raw>(`/company/jobs/${jobId}/close`, { method: "POST" }).then(normalizeJob),

  draftDescription: (input: {
    brief: string;
    title?: string;
    domain?: string;
    experienceLevel?: string;
    location?: string;
    skills?: string[];
  }): Promise<string> =>
    request<{ description: string }>("/company/jobs/draft-description", {
      method: "POST",
      body: {
        brief: input.brief,
        title: input.title,
        domain: input.domain,
        experience_level: input.experienceLevel,
        location: input.location,
        skills: input.skills ?? [],
      },
    }).then((r) => r.description),

  listColleges: (): Promise<CollegeOption[]> =>
    request<Raw[]>("/company/jobs/options/colleges").then((rows) =>
      rows.map((r) => ({
        id: r.id as string,
        name: r.name as string,
        city: (r.city as string | null) ?? null,
        state: (r.state as string | null) ?? null,
      })),
    ),

  listApplicants: (
    jobId: string,
    sortBy: "placement_probability" | "weighted_composite" = "placement_probability",
  ): Promise<RankedApplicant[]> =>
    request<Raw[]>(`/company/jobs/${jobId}/applicants?sort_by=${sortBy}`).then((rows) =>
      rows.map(normalizeRanked),
    ),

  getApplicantAnalysis: (jobId: string, applicationId: string): Promise<ApplicationAnalysis> =>
    request<Raw>(`/company/jobs/${jobId}/applicants/${applicationId}`).then(normalizeAnalysis),

  updateApplicantStatus: (
    jobId: string,
    applicationId: string,
    status: ApplicationStatus,
  ): Promise<Application> =>
    request<Raw>(`/company/jobs/${jobId}/applicants/${applicationId}/status`, {
      method: "PATCH",
      body: { status },
    }).then(normalizeApplication),
};

File: backend/app/routers/company/jobs.py
"""Jobs router — /company/jobs

The company side of the core platform loop: create a JD (optionally
Gemini-assisted), set per-job weight sliders, publish it Live. Distinct
from the legacy /company/roles feature.

All routes require a valid Bearer token AND role == "company", and every
job is scoped to the caller's own company. Uses the service-role db_client;
RLS in db/jobs_and_applications_migration.sql is the defense-in-depth layer.
"""

from __future__ import annotations

import logging
from datetime import date
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from postgrest.exceptions import APIError

from ...deps import require_company_role
from ...schemas import (
    DraftJDIn,
    DraftJDOut,
    JobCreateIn,
    JobOut,
    JobUpdateIn,
    JOB_STATUSES,
)
from ...crud import (
    count_applications_for_job,
    create_job,
    get_company_by_owner_id,
    get_job,
    get_job_skill_names,
    get_job_visible_college_ids,
    get_job_weights,
    list_jobs_by_company,
    resolve_skill_ids,
    set_job_skills,
    set_job_visible_colleges,
    update_job,
    upsert_job_weights,
)
from ...services.candidate.job_scoring_agent import draft_job_description
from ...deps import db_client

log = logging.getLogger(__name__)
router = APIRouter(prefix="/company/jobs", tags=["jobs"])


# ── Helpers ────────────────────────────────────────────────────────────

def _get_owned_company(current_user: dict) -> dict:
    company = get_company_by_owner_id(current_user["id"])
    if not company:
        raise HTTPException(
            status_code=404,
            detail="Company profile not found. Please complete onboarding.",
        )
    return company


def _get_owned_job(job_id: str, company_id: str) -> dict:
    job = get_job(job_id)
    if not job or job["company_id"] != company_id:
        raise HTTPException(status_code=404, detail="Job not found.")
    return job


def _job_out(job: dict) -> JobOut:
    weights = get_job_weights(job["id"])
    return JobOut(
        id=job["id"],
        company_id=job["company_id"],
        title=job["title"],
        description=job["description"],
        domain=job["domain"],
        experience_level=job["experience_level"],
        location=job["location"],
        openings_count=int(job.get("openings_count", 1)),
        deadline=str(job["deadline"]),
        visibility=job["visibility"],
        status=job["status"],
        skills=get_job_skill_names(job["id"]),
        visible_college_ids=get_job_visible_college_ids(job["id"]),
        weights={
            "resume_weight": weights["resume_weight"],
            "github_weight": weights["github_weight"],
            "leetcode_weight": weights["leetcode_weight"],
            "interview_weight": weights["interview_weight"],
            "assessment_weight": weights["assessment_weight"],
        } if weights else None,
        application_count=count_applications_for_job(job["id"]),
        employment_type=job.get("employment_type", "full-time"),
        interview_duration_minutes=int(job.get("interview_duration_minutes", 30)),
        interview_mode=job.get("interview_mode", "ai"),
        is_on_campus=bool(job.get("is_on_campus", True)),
        min_cgpa=job.get("min_cgpa"),
        eligible_branches=job.get("eligible_branches") or [],
        eligible_batch_years=job.get("eligible_batch_years") or [],
        ctc_min=job.get("ctc_min"),
        ctc_max=job.get("ctc_max"),
        ctc_currency=job.get("ctc_currency") or "INR",
        stipend_min=job.get("stipend_min"),
        stipend_max=job.get("stipend_max"),
        internship_duration_months=job.get("internship_duration_months"),
        ppo_ctc_min=job.get("ppo_ctc_min"),
        ppo_ctc_max=job.get("ppo_ctc_max"),
        perks=job.get("perks") or [],
        created_at=str(job.get("created_at", "")),
        updated_at=str(job.get("updated_at", "")),
    )


# ── POST /company/jobs/draft-description  (Gemini-assisted JD draft) ────

@router.post("/draft-description", response_model=DraftJDOut)
async def draft_description_route(
    payload: DraftJDIn,
    current_user: dict = Depends(require_company_role),
) -> DraftJDOut:
    _get_owned_company(current_user)
    try:
        text = await draft_job_description(
            brief=payload.brief,
            title=payload.title,
            domain=payload.domain,
            experience_level=payload.experience_level,
            location=payload.location,
            skills=payload.skills,
        )
    except RuntimeError as exc:  # missing API key
        raise HTTPException(status_code=503, detail=str(exc))
    except Exception as exc:  # noqa: BLE001
        log.exception("JD draft failed")
        raise HTTPException(status_code=502, detail=f"Could not draft a description: {exc}")
    return DraftJDOut(description=text)


# ── GET /company/jobs/options/colleges  (for the restricted-visibility picker) ─

@router.get("/options/colleges")
def list_colleges_route(
    current_user: dict = Depends(require_company_role),
) -> list:
    _get_owned_company(current_user)
    try:
        res = (
            db_client.table("colleges")
            .select("id, name, city, state")
            .order("name", desc=False)
            .execute()
        )
    except APIError as exc:
        log.error("DB error listing colleges: %s", exc)
        raise HTTPException(status_code=500, detail="Failed to fetch colleges.")
    return res.data or []


# ── POST /company/jobs  (create) ──────────────────────────────────────

@router.post("", response_model=JobOut, status_code=201)
def create_job_route(
    payload: JobCreateIn,
    current_user: dict = Depends(require_company_role),
) -> JobOut:
    company = _get_owned_company(current_user)

    if payload.publish and payload.deadline < date.today():
        raise HTTPException(
            status_code=400,
            detail="Cannot publish a job with a deadline in the past.",
        )

    try:
        skill_ids, _ = resolve_skill_ids(payload.skills)
        job = create_job(company["id"], {
            "title": payload.title,
            "description": payload.description,
            "domain": payload.domain,
            "experience_level": payload.experience_level,
            "location": payload.location,
            "openings_count": payload.openings_count,
            "deadline": payload.deadline.isoformat(),
            "visibility": payload.visibility,
            "status": "live" if payload.publish else "draft",
            # Eligibility & offer detail (jobs_eligibility_migration.sql).
            "employment_type": payload.employment_type,
            "interview_duration_minutes": payload.interview_duration_minutes,
            "interview_mode": payload.interview_mode,
            "is_on_campus": payload.is_on_campus,
            "min_cgpa": payload.min_cgpa,
            "eligible_branches": payload.eligible_branches,
            "eligible_batch_years": payload.eligible_batch_years,
            "ctc_min": payload.ctc_min,
            "ctc_max": payload.ctc_max,
            "ctc_currency": payload.ctc_currency,
            "stipend_min": payload.stipend_min,
            "stipend_max": payload.stipend_max,
            "internship_duration_months": payload.internship_duration_months,
            "ppo_ctc_min": payload.ppo_ctc_min,
            "ppo_ctc_max": payload.ppo_ctc_max,
            "perks": payload.perks,
        })
        if not job:
            raise HTTPException(status_code=400, detail="Failed to create job.")

        set_job_skills(job["id"], skill_ids)
        set_job_visible_colleges(
            job["id"],
            payload.visible_college_ids if payload.visibility == "restricted" else [],
        )
        # DB CHECK (job_weights_sum_100) re-validates the sum server-side.
        upsert_job_weights(job["id"], payload.weights.model_dump())
    except APIError as exc:
        log.error("DB error creating job for company %s: %s", company["id"], exc)
        raise HTTPException(status_code=400, detail=f"Failed to create job: {exc.message}")

    return _job_out(get_job(job["id"]))


# ── GET /company/jobs  (list) ─────────────────────────────────────────

@router.get("", response_model=List[JobOut])
def list_jobs_route(
    status: Optional[str] = None,
    current_user: dict = Depends(require_company_role),
) -> List[JobOut]:
    if status is not None and status not in JOB_STATUSES:
        raise HTTPException(status_code=400, detail=f"status must be one of {JOB_STATUSES}.")
    company = _get_owned_company(current_user)
    try:
        rows = list_jobs_by_company(company["id"], status)
    except APIError as exc:
        log.error("DB error listing jobs for company %s: %s", company["id"], exc)
        raise HTTPException(status_code=500, detail="Failed to fetch jobs.")
    return [_job_out(r) for r in rows]


# ── GET /company/jobs/{job_id} ───────────────────────────────────────

@router.get("/{job_id}", response_model=JobOut)
def get_job_route(
    job_id: str,
    current_user: dict = Depends(require_company_role),
) -> JobOut:
    company = _get_owned_company(current_user)
    return _job_out(_get_owned_job(job_id, company["id"]))


# ── PATCH /company/jobs/{job_id} ─────────────────────────────────────

@router.patch("/{job_id}", response_model=JobOut)
def update_job_route(
    job_id: str,
    payload: JobUpdateIn,
    current_user: dict = Depends(require_company_role),
) -> JobOut:
    company = _get_owned_company(current_user)
    job = _get_owned_job(job_id, company["id"])

    fields: dict = {}
    for attr in (
        "title", "description", "domain", "experience_level", "location",
        "openings_count", "visibility",
        # Eligibility & offer detail (jobs_eligibility_migration.sql).
        "employment_type", "interview_duration_minutes", "interview_mode", "is_on_campus",
        "min_cgpa", "eligible_branches", "eligible_batch_years",
        "ctc_min", "ctc_max", "ctc_currency",
        "stipend_min", "stipend_max", "internship_duration_months",
        "ppo_ctc_min", "ppo_ctc_max", "perks",
    ):
        val = getattr(payload, attr)
        if val is not None:
            fields[attr] = val
    if payload.deadline is not None:
        fields["deadline"] = payload.deadline.isoformat()

    new_visibility = payload.visibility or job["visibility"]
    if new_visibility == "restricted" and payload.visible_college_ids is not None and not payload.visible_college_ids:
        raise HTTPException(status_code=400, detail="Restricted jobs must list at least one college.")

    try:
        if fields:
            update_job(job_id, fields)
        if payload.skills is not None:
            skill_ids, _ = resolve_skill_ids(payload.skills)
            set_job_skills(job_id, skill_ids)
        if payload.visible_college_ids is not None or payload.visibility is not None:
            set_job_visible_colleges(
                job_id,
                (payload.visible_college_ids or get_job_visible_college_ids(job_id))
                if new_visibility == "restricted" else [],
            )
        if payload.weights is not None:
            upsert_job_weights(job_id, payload.weights.model_dump())
    except APIError as exc:
        log.error("DB error updating job %s: %s", job_id, exc)
        raise HTTPException(status_code=400, detail=f"Update failed: {exc.message}")

    return _job_out(get_job(job_id))


# ── POST /company/jobs/{job_id}/publish ─────────────────────────────

@router.post("/{job_id}/publish", response_model=JobOut)
def publish_job_route(
    job_id: str,
    current_user: dict = Depends(require_company_role),
) -> JobOut:
    company = _get_owned_company(current_user)
    job = _get_owned_job(job_id, company["id"])

    if date.fromisoformat(str(job["deadline"])) < date.today():
        raise HTTPException(status_code=400, detail="Cannot publish a job with a past deadline.")
    if not get_job_weights(job_id):
        raise HTTPException(status_code=400, detail="Set the weight sliders before publishing.")

    update_job(job_id, {"status": "live"})
    return _job_out(get_job(job_id))


# ── POST /company/jobs/{job_id}/close ───────────────────────────────

@router.post("/{job_id}/close", response_model=JobOut)
def close_job_route(
    job_id: str,
    current_user: dict = Depends(require_company_role),
) -> JobOut:
    company = _get_owned_company(current_user)
    _get_owned_job(job_id, company["id"])
    update_job(job_id, {"status": "closed"})
    return _job_out(get_job(job_id))
